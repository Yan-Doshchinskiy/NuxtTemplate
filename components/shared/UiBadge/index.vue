<template>
  <span
    class="ui-badge"
    :class="badgeClass"
  >
    {{ label }}
  </span>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator';

import { COMPUTED_STYLE } from '~/enums/ui';

export type TBadgeSize = 'medium' | 'small';
export type TBadgeVariant = 'default' | 'warning' | 'danger' | 'success' | 'info';

@Component({
  name: 'UiBadge'
})
export default class UiBadge extends Vue {
  @Prop({ default: '' }) readonly label!: string;
  @Prop({ default: 'default' }) readonly variant!: TBadgeVariant;
  @Prop({ default: 'medium' }) readonly size!: TBadgeSize;

  private get badgeClass(): COMPUTED_STYLE {
    const {
      variant,
      size
    } = this;
    return [
      {
        'ui-badge_default': variant === 'default',
        'ui-badge_warning': variant === 'warning',
        'ui-badge_danger': variant === 'danger',
        'ui-badge_success': variant === 'success',
        'ui-badge_info': variant === 'info',
        'ui-badge_medium': size === 'medium',
        'ui-badge_small': size === 'small'
      }
    ];
  }
}
</script>

<style lang="scss" scoped>

.ui-badge {
  height: max-content;
  width: max-content;

  display: flex;
  align-items: center;
  justify-content: center;

  border-radius: 4px;

  cursor: default;

  // color modifiers
  &_default {
    background-color: $gray-200;
    color: $black;
  }

  &_warning {
    background-color: transparentize($brown, $transparentize);
    color: $brown;
  }

  &_danger {
    background-color: transparentize($red, $transparentize);
    color: $red;
  }

  &_success {
    background-color: transparentize($teal, $transparentize);
    color: $teal;
  }

   //special color and size modifier
  &_info {
    background-color: $brown-light;
    color: $black;
    padding: 0 4px;
    min-height: 20px;
  }

  // size modifiers
  &_medium {
    padding: 4px 8px;
    font-weight: 400;
    font-size: 14px;
    line-height: 18px;
    min-height: 28px;
  }

  &_small {
    padding: 2px 4px;
    font-weight: 700;
    font-size: 10px;
    line-height: 10px;
    min-height: 14px;
  }

}
</style>
