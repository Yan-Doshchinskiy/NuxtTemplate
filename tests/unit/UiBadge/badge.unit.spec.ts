import UiBadge from '~/components/shared/UiBadge/index.vue';

import MainUnit from '~/tests/unit/prototypes/main.unit';
import MainMounterUtil from '~/tests/unit/prototypes/utils/main.mounter.util';
import MainOptionsUtil from '~/tests/unit/prototypes/utils/main.options.util';
import BadgeStylerUtil from '~/tests/unit/UiBadge/utils/badge.styler.util';
import BadgePropsPart from '~/tests/unit/UiBadge/parts/badge.props.part';
import BadgeStylePart from '~/tests/unit/UiBadge/parts/badge.style.part';

import type { IWrapperOptions } from '~/tests/unit/prototypes/utils/main.options.util';
import type { Slots } from '@vue/test-utils';
import type { TBadgeSize, TBadgeVariant } from '~/components/shared/UiBadge/index.vue';

export interface IComponentDefaultProps {
  label: string,
  variant: TBadgeVariant,
  size: TBadgeSize
}

export interface IComponentProps extends Partial<IComponentDefaultProps> {

}

export interface IComponentSlots extends Slots{

}

const defaultComponentProps: IComponentDefaultProps = {
  label: '',
  variant: 'default',
  size: 'medium'
};

const defaultTestOptions: IWrapperOptions<IComponentProps, IComponentSlots> = {
  propsData: {
  }
};

export type TOptionUtil = MainOptionsUtil<IComponentDefaultProps, IComponentProps, IComponentSlots>;
export type TStylerUtil = BadgeStylerUtil;
export type TMounterUtil = MainMounterUtil<IComponentDefaultProps, IComponentProps, IComponentSlots>;

class BadgeUnitSpec extends MainUnit<IComponentDefaultProps, IComponentProps, IComponentSlots, TOptionUtil, TStylerUtil, TMounterUtil> {
  constructor() {
    const optionUtil = new MainOptionsUtil({
      defaultComponentProps,
      defaultTestOptions
    });
    const stylerUtil = new BadgeStylerUtil({ wrapperClass: 'ui-badge' });
    const mounterUtil = new MainMounterUtil({
      component: UiBadge,
      wrapperClass: stylerUtil.wrapperClass,
      defaultTestOptions: optionUtil.defaultTestOptions
    });

    super({
      optionUtil,
      stylerUtil,
      mounterUtil
    });
  }
}

const testInstance = new BadgeUnitSpec();

testInstance.addNewTestPart((utils) => new BadgePropsPart(utils));
testInstance.addNewTestPart((utils) => new BadgeStylePart(utils));

testInstance.run();

export default BadgeUnitSpec;
