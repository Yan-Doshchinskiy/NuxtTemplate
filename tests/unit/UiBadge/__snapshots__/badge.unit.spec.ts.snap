// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UiBadge props works correctly with default props 1`] = `
<span
  class="ui-badge ui-badge_default ui-badge_medium"
>
  
  

</span>
`;

exports[`UiBadge style should display correct class with different variants and sizes 1`] = `
<span
  class="ui-badge ui-badge_default ui-badge_medium"
>
  
  

</span>
`;

exports[`UiBadge style should display correct class with different variants and sizes 2`] = `
<span
  class="ui-badge ui-badge_default ui-badge_small"
>
  
  

</span>
`;

exports[`UiBadge style should display correct class with different variants and sizes 3`] = `
<span
  class="ui-badge ui-badge_warning ui-badge_medium"
>
  
  

</span>
`;

exports[`UiBadge style should display correct class with different variants and sizes 4`] = `
<span
  class="ui-badge ui-badge_warning ui-badge_small"
>
  
  

</span>
`;

exports[`UiBadge style should display correct class with different variants and sizes 5`] = `
<span
  class="ui-badge ui-badge_danger ui-badge_medium"
>
  
  

</span>
`;

exports[`UiBadge style should display correct class with different variants and sizes 6`] = `
<span
  class="ui-badge ui-badge_danger ui-badge_small"
>
  
  

</span>
`;

exports[`UiBadge style should display correct class with different variants and sizes 7`] = `
<span
  class="ui-badge ui-badge_success ui-badge_medium"
>
  
  

</span>
`;

exports[`UiBadge style should display correct class with different variants and sizes 8`] = `
<span
  class="ui-badge ui-badge_success ui-badge_small"
>
  
  

</span>
`;

exports[`UiBadge style should display correct class with different variants and sizes 9`] = `
<span
  class="ui-badge ui-badge_info ui-badge_medium"
>
  
  

</span>
`;

exports[`UiBadge style should display correct class with different variants and sizes 10`] = `
<span
  class="ui-badge ui-badge_info ui-badge_small"
>
  
  

</span>
`;
